File: app/src/main/java/com/mulechat/app/identity/Identity.java
package com.mulechat.app.identity;

import java.util.Arrays;

/**
 * A user's self-sovereign cryptographic identity (Ed25519 keypair + a
 * short peerId derived from the public key). The private key never leaves
 * the device unencrypted — see storage/ for how it's persisted.
 */
public final class Identity {
    public final byte[] publicKey;
    public final byte[] privateKey;
    public final String peerId;

    public Identity(byte[] publicKey, byte[] privateKey, String peerId) {
        this.publicKey = publicKey;
        this.privateKey = privateKey;
        this.peerId = peerId;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Identity)) return false;
        Identity other = (Identity) o;
        return Arrays.equals(publicKey, other.publicKey)
                && Arrays.equals(privateKey, other.privateKey)
                && peerId.equals(other.peerId);
    }

    @Override
    public int hashCode() {
        int result = Arrays.hashCode(publicKey);
        result = 31 * result + Arrays.hashCode(privateKey);
        result = 31 * result + peerId.hashCode();
        return result;
    }
}

File: composeApp/src/commonMain/kotlin/com/nulchat/identity/Identity.kt
package com.nulchat.identity

/**
 * A user's self-sovereign cryptographic identity.
 *
 * - [publicKey] / [privateKey] are raw Ed25519 key bytes (32 bytes each).
 * - [peerId] is a human-shareable identifier derived from the public key
 *   (Blake2b hash, base32-encoded — similar in spirit to libp2p peer IDs).
 *
 * The private key never leaves the device unencrypted; it is stored only
 * inside the SQLCipher-encrypted local database (see storage/).
 */
data class Identity(
    val publicKey: ByteArray,
    val privateKey: ByteArray,
    val peerId: String
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Identity) return false
        return publicKey.contentEquals(other.publicKey) &&
            privateKey.contentEquals(other.privateKey) &&
            peerId == other.peerId
    }

    override fun hashCode(): Int {
        var result = publicKey.contentHashCode()
        result = 31 * result + privateKey.contentHashCode()
        result = 31 * result + peerId.hashCode()
        return result
    }
}
